mmap-alloc/src/lib.rs

```diff
@@ -306,59 +306,64 @@ impl MapAlloc {
+        // Finally, we tell the OS that the page at NULL is unused, but do not unmap it.
+        // Since we leave the page at 0 mapped, future calls should never return null.
         // Note that this leaks memory since we never unmap that page, but this isn't a big deal -
         // since we never write to it, it will remain uncommitted and will thus not consume any
         // physical memory.
         #[cold]
-        unsafe fn fix_null(allocator: &MapAlloc, size: usize) ->  *mut u8 {
+        unsafe fn fix_null(allocator: &MapAlloc, size: usize, first_byte: u8) ->  *mut u8 {
             // First create a mapping that will serve as the destination of the remap
-            let new_ptr = map(size, allocator.perms, false, allocator.huge_pagesize)
+            let new_ptr = map(size, perms::PROT_WRITE, false, allocator.huge_pagesize)
                 .expect("Not enough virtual memory to make new mapping");
             debug_assert!(!new_ptr.is_null(),
                           "we have an open mapping on null, new mapping should never be null");
 
             // remap onto the newly-created mapping. This should never fail because the target
             // mapping already exists.
-            let move_result = libc::mremap(
-                ptr::null_mut(),
-                size,
-                size,
-                libc::MREMAP_MAYMOVE | libc::MREMAP_FIXED,
-                new_ptr
-            );
-            // Should never fail: we're remapping into an existing mapping
-            assert_ne!(move_result, libc::MAP_FAILED, "Unable to move mapping: {}", errno());
-
-            // Map a singe page at 0 to ensure this doesn't happen again
-            let pagesize = sysconf::page::pagesize();
-            // TODO: Race condition if another mmap/mremap returned null after we mremaped, but before mmap
-            let null_map = libc::mmap(
-                ptr::null_mut(),
-                pagesize,
-                libc::PROT_NONE,
-                libc::MAP_ANONYMOUS | libc::MAP_FIXED | libc::MAP_NORESERVE,
-                -1,
-                0
-            );
-            // Ignore errors creating the new map at 0.
-            // This should never fail
-            if null_map.is_null() {
-                // a) Make it more likely that the kernel will not keep the page backed by physical
-                // memory and, b) make it so that an access to that range will result in a segfault to
-                // make other bugs easier to detect.
-                mark_unused(ptr::null_mut(), pagesize);
+            let move_size = size - allocator.pagesize;
+            if move_size > 0 {
+                let move_result = libc::mremap(
+                    allocator.pagesize as *mut _,
+                    size,
+                    size,
+                    libc::MREMAP_MAYMOVE | libc::MREMAP_FIXED,
+                    new_ptr.offset(allocator.pagesize as isize)
+                );
```

Comment:
Shouldn't the old size here be `size - allocator.pagesize` since we're not moving the first page?

Reply:
Also, why are we specifying `MREMAP_FIXED`? I would assume that we don't care where the mapping is moved to. What if, for example, our allocation is 2 pages long, and page 2 is already mapped? Then this will just fail.

Reply:
You're right, both `size`s should be `move_size`.

Reply:
Oh I think the old size should be `move_size` and the new size should be `size`.

Reply:
That way the final mapping is the right size.

Reply:
We want the mapping moved on top of the new mapping we made (offset by 1 page). If we let it choose anywhere, there would be no guarentee that there was an available page 1 page beforehand. Also, `mremap(p, s, s, ..)` without `MREMAP_FIXED` should be a no-op: there's no need to move the mapping if it's not growing.

Reply:
OH sorry I totally missed what you were doing here, nvm. For some reason I thought you'd abandoned the old mechanism of pre-mapping the new allocation before doing `mremap`. I see what's going on here.